text/0000-dockerfiles.md

```diff
@@ -0,0 +1,228 @@
+
+# Summary
+[summary]: #summary
+
+This RFC introduces functionality for customizing base images, as an alternative to stackpacks.
+
+# Motivation
+[motivation]: #motivation
+
+Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with build-time and runtime Dockerfiles that act as pre-build hooks.
+These hooks would be equivalent to base image / builder customizations, but would execute immediately before a build.
+
+Most importantly: the changes below do not impact buildpacks or the buildpack API.
+For a given application, a build that uses hooks could be optimized by creating a more narrowly-scoped builder with the hooks pre-applied, without modifying the buildpacks.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
+
+### Dynamically-applied Dockerfiles
+
+A builder image may include any number of "hook" files in `/cnb/hooks.d/`. Hooks may be associated with zero or more buildpack IDs.
+The `pack create-builder` command is used to copy hooks into the builder image via `builder.toml`. E.g.,
+```toml
+[[hooks]]
+name = "app-hook"
+path = "./myhook"
+buildpacks = ["com.example.buildpack1"]
+```
+
+(Note: this format should not be considered final. Additional fields will be required to mark the format, target, etc. of the hook. These may be defined in subsequent spec PRs / sub-team RFCs.)
+
+Each hook file path must be in the format `/cnb/hooks.d/<name>.(build.|run.|)<format>(.out|)`, where:
+
+1. `build`, `run`, or empty specify the builder image, runtime base image, or both base images, respectively.
+2. The only valid format is `Dockerfile`, although support for, e.g. LLB JSON, could be added in the future.
+3. If the `.out` suffix is present and the file is executable, the hook is executed in the context of the app directory, and its output must match `<format>`.
+4. If the `.out` suffix is not present, the contents of the file must match `<format>`.
+
+Additionally, each hook file may be accompanied by an additional file of the same name, but with the suffix `.buildpacks`.
+This file contains a newline-separated list of buildpack IDs.
+If present, the hook is ignored unless one of the buildpacks in the list is selected during the buildpack detection phase.
+
+Hook files are evaluated (either read or executed) after the detection phase.
+All Dockerfiles are applied to base images after the hooks are evaluated and before the buildpack build phase.
+The resulting Dockerfiles are applied to the build-time and runtime base images in lexographical order by `<name>`.
+
+Executable hook files must not modify the app directory when executed and may be evaluated in parallel.
```

Comment:
@sclevine Most Paketo buildpacks use env variables to receive configuration from users. Many of these settings are not likely useful to hooks, but we do allow users to specify an env variable to pick versions of dependencies installed, ex: `BP_JVM_VERSION=8`. 

Will hooks have the same set of env variables set that you'd get during detect & build?